Migrate API reference documentation to MkDocs

gordonhart opened the pull request:
This PR migrates this project's docs engine to [MkDocs](https://www.mkdocs.org/) from Sphinx. It uses the [mkdocs-material](https://squidfunk.github.io/mkdocs-material/) theme and [mkdocstrings](https://github.com/mkdocstrings/mkdocstrings) to extract docstrings from source code. There are many knobs and levers to tune in `mkdocs.yml` and in putting together this PR I've explored most/all of them and am quite happy with the result.

The diff is larger than I'd like, mainly due to subtly different interpretation of docstrings:

- We're still using the [Sphinx docstring style](https://mkdocstrings.github.io/python/usage/docstrings/sphinx/), but markup is now in Markdown instead of reST
- References are now formatted ```[`ClassName`][path.to.ClassName]``` instead of ```:class:`path.to.ClassName` ```
- Instead of accepting attribute docstrings as `#:` comments in front of a docstring, mkdocstrings wants `"""string"""` docstrings _following_ the attribute definition
- [mkdocstrings doesn't support inheritance](https://github.com/mkdocstrings/mkdocstrings/issues/78#issuecomment-1511013734), which means there's a fair bit of boilerplate necessary to get legacy `classification` and `detection` definitions properly documented 

The first of many PRs in this chain. Planning to follow up with PRs to customize styles, migrate [docs.kolena.io](https://docs.kolena.io/) content into this repo, set up automated publishing, and add metrics guides (among other things).

Fixes KOL-2457
Fixes KOL-2459

gordonhart committed:
Update dependencies to mkdocs only

gordonhart committed:
Start wiring together mkdocs w/ mkdocstrings -- looks promising so far

gordonhart committed:
Continue tuning theme, update initialize documentation

gordonhart committed:
Use default themes to unblock

gordonhart committed:
Use Sphinx docstring style; add social footer links

gordonhart committed:
Start wiring together reference documentation

gordonhart committed:
Wire together kolena.workflow reference docs

gordonhart committed:
Continue porting reference documentation

gordonhart committed:
Cleanliness pass through annotations, assets

gordonhart committed:
Continue cleaning up kolena.workflow API docs; start porting kolena.classification into legacy section

gordonhart committed:
Document kolena.fr

gordonhart committed:
Continue fixing up kolena.classification legacy docs

gordonhart committed:
Simplify built-in kolena.fr setup

gordonhart committed:
Add indentation to docs blocks for legibility

gordonhart committed:
Add docstrings to various plot definitions; move define_workflow into a public module

gordonhart committed:
Improve API reference documentation -- main hangup is still missing inheritance from kolena.{classification,detection}

gordonhart committed:
Merge branch 'trunk' into gh/mkdocs

gordonhart committed:
Add kolena.workflow.metrics reference docs

gordonhart committed:
Tune up kolena.detection docstrings

gordonhart committed:
Touch up kolena.detection._internal docs -- still not committed to a pattern for inherited members

gordonhart committed:
Reference base classes directly; flip strict mode back on

gordonhart committed:
Include base class documentation and reference where applicable

gordonhart committed:
Start cleaning up in preparation for review

gordonhart committed:
Add CI step to build docs on Python 3.9 only

gordonhart committed:
Merge branch 'trunk' into gh/mkdocs

gordonhart committed:
Update docs build and publish to use GH Pages instead of S3

gordonhart committed:
Remove working directory from docs build step

gordonhart committed:
Stub out pages; use icon instead of wordmark

gordonhart committed:
Cleanup in preparation for review

gordonhart commented on CONTRIBUTING.md:
The plan is to include developer documentation in this file. "How to contribute to this repo." The root README is will be targeted more at users planning to use `kolena`, rather than those planning to submit PRs.

gordonhart commented on docs/index.md:
Not decided yet how closely we want this to mirror the repo's root README. Planning to revisit this landing page when migrating over content from [docs.kolena.io](https://docs.kolena.io/) in a follow-up PR.

gordonhart committed:
Add basic usage snippets to the top of various files

gordonhart commented:
I'm happy with where this is for a first pass. Requesting review specifically from @ly-kolena, @munkyshi, and @y27choi — will probably be easiest to render locally with `poetry run mkdocs serve -a localhost:8999` and explore in your browser.

ly-kolena commented on .github/workflows/release.yml:
```diff
@@ -44,21 +44,10 @@ jobs:
-          aws s3 sync ./build "$BUCKET_PATH"
-        working-directory: ./docs
+          poetry run mkdocs build
+          poetry run mkdocs gh-deploy
```
Non-blocking note: Need to somehow protect this branch. I just tried to run this command and the branch got created.

gordonhart replied:
👍 will add protections. Want to run this path once in CI to see how the commits look, then restrict the branch to only accept that sort of commit.

ly-kolena commented on mkdocs.yml:
```diff
@@ -0,0 +1,129 @@
+repo_name: kolenaIO/kolena
+repo_url: https://github.com/kolenaIO/kolena
+copyright: Copyright &copy; 2021-2023 Kolena
+edit_uri: edit/trunk/docs/
```
Non-blocking nit: How do we expect this to work in terms of process?

gordonhart replied:
It will prompt users to fork the repo to submit a PR. I think it's nice to include to lower the barrier to entry for small contributions (e.g. fixing spelling or grammar) and a nice prod to remind readers that the project is open-source and accepts community PRs.

ly-kolena reviewed (approved):
Sorta miss some of the alphabetical sort from previous API layout, but it may just be a personal preference or habit.

gordonhart commented:
> Sorta miss some of the alphabetical sort from previous API layout

Glad that you flagged this. I decided to go with a "usage ordering" rather than alphabetical for `kolena.workflow` definitions under the assumption that it would be easier to browse. What do others think?

| "Usage Order" | Alphabetical |
| ------------- | ---------------- |
| <img width="264" alt="Screen Shot 2023-06-08 at 7 48 46 AM" src="https://github.com/kolenaIO/kolena/assets/9648886/ab5061e8-601e-47c9-9609-57a6de6b1508"> | <img width="233" alt="Screen Shot 2023-06-08 at 7 49 28 AM" src="https://github.com/kolenaIO/kolena/assets/9648886/2c5c6eca-7520-4f2a-a252-4603a65d4cb3"> |

gordonhart committed:
Expand on various kolena.workflow docstrings

gordonhart committed:
Expand on kolena.{errors,initialize} docstrings and add PyPI footer link

munkyshi commented on kolena/classification/model.py:
```diff
@@ -68,13 +73,15 @@ def _inferences_from_record(self, record: Any) -> Tuple[TestImage, Optional[List
     """
-    A :class:`kolena.classification.Model` with a special :meth:`kolena.classification.InferenceModel.infer` member
-    performing inference on a provided :class:`kolena.classification.TestImage`.
+    A [`Model`][kolena.classification.Model] with a special [`infer][kolena.classification.InferenceModel.infer] member
```
Closing backtick for `infer`?

munkyshi commented on kolena/classification/test_case.py:
```diff
@@ -33,13 +33,16 @@ class TestCase(BaseTestCase):
+        provided name does not already exist.
+    :param description: Optionally specify a description for the new test case. Ignored when a test case with the
+        provided name already exists.
+    :param images: Optionally provide a list of [TestImage][kolena.classification.TestImage] images used to seed a new
```
nit: Seems like we're wrapping classes in backticks elsewhere, so should probably do so for `TestImage`

munkyshi commented on kolena/detection/_internal/test_suite.py:
```diff
@@ -211,28 +212,28 @@ def description(self, description: str) -> None:
             If a different version of the test case already exists in this test suite, it is replaced.
 
-            :param test_case: the test case to add to the test suite
+            :param test_case: The test case to add to the test suite.
```
Stray thought -- given that you're going through the effort of making param structure consistently full sentences (caplitalized, with periods), should we specify or link something as a style guide in the `CONTRIBUTING.md`?

gordonhart replied:
Good idea. Will add this to the backlog — want to make sure that the style guide is complete before committing anything.

munkyshi reviewed (approved):
Skimmed the first 60 or so files, left some small comments, nothing blocking.

gordonhart committed:
Address PR comments

ly-kolena commented:
> > Sorta miss some of the alphabetical sort from previous API layout
> 
> Glad that you flagged this. I decided to go with a "usage ordering" rather than alphabetical for `kolena.workflow` definitions under the assumption that it would be easier to browse. What do others think?

Non-blocking footnote, the TOC of legacy workflows is a bit more jarring in terms of habits.

y27choi commented on kolena/workflow/test_sample.py:
`Image`, `ImagePair`, `Text`, `ImageText` are missing fields (i.e. `locator`, `a`, `b`, `text`, `image`)
<img width="624" alt="image" src="https://github.com/kolenaIO/kolena/assets/17771952/24f8524a-f7dd-4ff6-9359-aee823c7baed">

y27choi replied:
Also missing the full API (with the expected arguments) in this page. 

<img width="642" alt="image" src="https://github.com/kolenaIO/kolena/assets/17771952/4abb7199-64e3-4420-aa5a-6b8084b497a2">

gordonhart replied:
Nice catch! Documented.

gordonhart committed:
Catch a number of lagging Sphinx-style references

gordonhart committed:
Document fields of test sample types

gordonhart committed:
Order legacy built-in workflow members alphabetically

gordonhart commented:
> Non-blocking footnote, the TOC of legacy workflows is a bit more jarring in terms of habits.

Ack — updated these to be [mostly] alphabetical in the latest commit.

gordonhart committed:
Add missing  to kolena.workflow.Video docs

y27choi commented:
Do we want to add the source code block for every page or only the ones that would benefit from having?
I noticed that we have it for `TestCase` but not for `TestSample` or `GroundTruth`.

y27choi commented on kolena/workflow/test_case.py:
Should we add `return type` details to functions like `load_test_samples()` and `iter_test_samples()`? 

<img width="516" alt="image" src="https://github.com/kolenaIO/kolena/assets/17771952/da3a3131-4d89-417e-ab7b-2f5124898739">

<img width="652" alt="image" src="https://github.com/kolenaIO/kolena/assets/17771952/f3d51a24-3934-411c-8cec-9a06c16549b1">

gordonhart replied:
Were missing docstrings — added.

y27choi commented on kolena/workflow/test_case.py:
I'm not sure if there is a way to define the order but it would be nice to have ` Editor` show up last instead of in between the class definition and class methods.  It's not too obvious to me that class methods for `TestCase` belong to `TestCase` and not `Editor`.

gordonhart replied:
Agreed. We don't have great ability to control this, so it's going to be a rough edge for the foreseeable future.

y27choi commented on kolena/workflow/test_suite.py:
not sure if it was intentional but `load_test_samples()` is missing.  It exists in the previous docs but without any description.

gordonhart replied:
👍 it was missing a docstring. Added

gordonhart commented:
> Do we want to add the source code block for every page or only the ones that would benefit from having?

I'm going to disable this flag entirely. Source code is usually not relevant when viewing documentation, and the callouts are pretty distracting.

gordonhart committed:
Add a few missing docstrings; disable source code embedding

y27choi commented on kolena/workflow/evaluator.py:
```diff
@@ -337,13 +389,13 @@ def display_name(self) -> str:
+    Test-case-level plots ([`Plot`][kolena.workflow.Plot]) may also be computed.
     """
 
     configurations: List[EvaluatorConfiguration]
```
For class `Evaluator Configuration` and `Evaluator`, `display_name()` is missing in the new API doc.
Also, input arg `configurations` info is missing?

y27choi commented on kolena/workflow/asset.py:
Missing `locator` description on `BinaryAsset`

gordonhart committed:
Document missing methods on Evaluator and attributes in other places

y27choi commented on kolena/workflow/metrics/_geometry.py:
I'm wondering if we should specify what `GT`, `Inf`, `GT_Multiclass` and `Inf_Multiclass` are.  Maybe similar to how we surfaced `NumberSeries` and `NullabledNumberSeries` in `evaluator.py`

gordonhart replied:
Intentionally omitted these — should be clear from the function docstring and param snippets what can be accepted here.

y27choi commented on kolena/workflow/workflow.py:
```diff
@@ -90,22 +90,20 @@ def _from_api_response(cls, workflow: str, response: API.EvaluatorResponse) -> "
 
 @dataclass(frozen=True)
 class Workflow:
```
I think this class is missing from the doc (or I can't seem to easily find it).

gordonhart replied:
I decided to omit it — it's not a useful user-facing class, and it has a few methods on it that it isn't 100% safe to remove but also aren't a part of the "public" API (see #59).

y27choi replied:
Right. That's fair.

y27choi commented on docs/reference/built-in/classification.md:
```diff
@@ -0,0 +1,22 @@
+
+::: kolena.classification
+    options:
+      members_order: alphabetical
```
Isn't it better to order the attributes in the same order as we would expect in the API?
For example, `TestImage(locator, dataset=None, labels=None, metadata=None)`, I would expect to see the attirbutes in the following order: `locator`, `dataset`, `labels` and `metadata`.

<img width="626" alt="image" src="https://github.com/kolenaIO/kolena/assets/17771952/21098d7f-422d-4e92-bf46-83423a2fc81f">

gordonhart replied:
This members order applies to all classes as well as members within a class. Toggling this switch is a trade-off — better ordering in TOC, with worse ordering of these params here. Considering that it's just instance attributes that are impacted here, I think this tradeoff is worth it.

y27choi replied:
Hmm I see.  Well that's a bummer but I agree that it is worth it.

y27choi commented on docs/reference/built-in/classification.md:
```diff
@@ -0,0 +1,22 @@
+- [`kolena.classification.TestRun`][kolena.classification.TestRun]: test models on test suites
+- [`kolena.classification.Model`][kolena.classification.Model]: create models for testing
+
+::: kolena.classification
```
missing `kolena.classification.multiclass`

gordonhart replied:
Intentionally omitted to prevent this PR from growing even larger. The `kolena.classification.multiclass` implementation is missing a fair number of docstrings that still need to be written. Also haven't decided where builtin-generic workflows should live. Planning to get to that in a follow-up [KOL-2487](https://linear.app/kolena/issue/KOL-2487/reference-documentation-for-builtin-generic-workflows).

y27choi commented on kolena/fr/test_suite.py:
Also missing some class methods and `Editor`.

gordonhart replied:
Can you point to the exact methods that are missing? I see everything I would expect to see for `kolena.fr.TestSuite`.

y27choi replied:
Ah I thought I got rid of all the comments on this.  I didn't realize there was the Base Definitions section which covers all class methods and editor functions.

y27choi commented on kolena/detection/_internal/model.py:
```diff
@@ -55,15 +55,18 @@
 class BaseModel(ABC, Frozen, WithTelemetry):
     """
-    The descriptor for your model within the Kolena platform.
+    The base class for [`kolena.classification.Model`][kolena.classification.Model] and
```
I wonder if we should explicitly mention `kolena.classification.InferenceModel` here as well but given that it's inherited from `kolena.classification.Model` maybe it's not necessary.

y27choi commented on kolena/detection/_internal/test_suite.py:
```diff
@@ -211,42 +212,42 @@ def description(self, description: str) -> None:
             self._test_cases.pop(name)
 
-        @deprecated(details="use :meth:`add` instead", deprecated_in="0.56.0")
+        @deprecated(details="use `TestSuite.Editor.add`", deprecated_in="0.56.0")
```
This deprecate warning doesn't seem to show up on the[ doc](http://localhost:8999/reference/built-in/base/)

y27choi replied:
Same thing on fr.TestSuite.merge as well. [link to doc](http://localhost:8999/reference/built-in/fr/#kolena.fr.test_suite.TestSuite.Editor.merge)

gordonhart replied:
Thanks for flagging — the previous Sphinx setup had a good integration with the `deprecation` library and this new mkdocstrings setup does not. Taking a pass through all uses of `@deprecated` to ensure that the callout has been added to the docstring.

y27choi commented:
There are duplicate `GroundTruth` and `Inference` sections on `kolena.detection` page.  Should we remove the `GroundTruth` and `Inference` reference at the top?
<img width="198" alt="Screen Shot 2023-06-08 at 12 20 31 PM" src="https://github.com/kolenaIO/kolena/assets/17771952/e133c89f-19a4-43a9-88bc-5340ff912c0d">

y27choi commented:
Finished the review and left some comments.  Mainly compared the new docs with the old api docs and flagged the differences.

gordonhart committed:
Add a few missing deprecation callouts

gordonhart commented:
Thanks for the detailed review @y27choi — I believe I've addressed all of your points that fit into the scope of this PR. Tracking the rest in follow-ups.
